pydm/utilities/__init__.py

```diff
@@ -173,3 +176,64 @@ def _access_check(fn, mode):
                 if _access_check(name, mode):
                     return name
     return None
+
+
+def nested_dict_get(input_dict, nested_key):
+    internal_dict_value = input_dict
+    for k in nested_key:
+        try:
+            result = list(filter(None, re.split("(\[.+])", k)))
+            if len(result) > 1:
+                internal_dict_value = internal_dict_value.get(result[0], None)
+                internal_dict_value = eval(
+                    "internal_dict_value{}".format(''.join(result[1:]))
+                )
+            else:
+                internal_dict_value = internal_dict_value.get(k, None)
+        except:
+            internal_dict_value = None
+        if internal_dict_value is None:
+            return None
+    return internal_dict_value
+
+
+def data_callback(widget, data, introspection, mapping):
+    """
+    This callback executes the methods mapped at `mapping` with the data
+    from the channel following the introspection definition.
+
+    Parameters
+    ----------
+    widget : QWidget
+        The widget being affected by this callback
+    data : dict
+        Data from the channel
+    introspection : dict
+        Mapping between DataKey and plugin data fields
+    mapping : dict
+        Map containing the relation between DataKey and method for widget
+    """
+    if data is None:
+        return
+    try:
+        for data_key, real_key in introspection.items():
+            if real_key is None or real_key == '' or not isinstance(real_key, str):
+                continue
+            try:
+                method_name = mapping[data_key]
+            except KeyError:
+                continue
+            if isinstance(method_name, str):
+                method = getattr(widget, method_name, None)
+            elif callable(method_name):
+                method = method_name
+            else:
+                method = None
+            if not method:
+                continue
+            new_value = nested_dict_get(data, real_key.split('.'))
+            if new_value is not None:
+                method(new_value)
+    except RuntimeError:
```

Comment:
Should consider making a special pydm exception here (as I assume pydm raises `RuntimeError` somehow when the widget is being destroyed, based on the comment below?)